### PR TITLE
[Retry] Ignore outdated retry pods in monitor to prevent scheduler race

### PR DESCRIPTION
**Pull request by Yacouby**
### 📝 Description
This PR fixes a race condition in the runs monitor where pods from previous retry attempts were incorrectly considered when updating the run state.
The issue could cause the scheduler to start overlapping runs because the DB temporarily marked a retried run as error.

Now, the monitor ignores outdated pods and only tracks the latest retry attempt, ensuring correct run state and preventing duplicate runs.

---

### 🛠️ Changes Made

- Added `_is_pod_from_outdated_retry` to detect pods from outdated retry attempts.
- Updated monitor logic to skip outdated pods when determining run state.
- Improved debug logging for better visibility when updating a run's state.

---

### ✅ Checklist
- [ ] I updated the documentation (if applicable)
- [x] I have tested the changes in this PR

---

### 🧪 Testing

- Added a unit test to test the new functionality added.
- Manual validation on patched vm showed that the scheduler no longer triggers overlapping runs for jobs with retries.
- Modified an existing unit test (`test_monitor_run_retry_exhausted`) to label pods with their retry attempt, since unlabeled pods would otherwise be skipped by the monitor logic and fail that test.

---

### 🔗 References
- Ticket link: https://iguazio.atlassian.net/browse/ML-10803

---

### 🚨 Breaking Changes?

- [ ] Yes (explain below)
- [x] No

### CODE REVIEW

**Review by weilerN — APPROVED** [review]

File: server/py/services/api/runtime_handlers/base.py [hunk 1]
```diff
@@ -1767,7 +1780,11 @@ def _ensure_run_state(
                     run, reason, message
                 )
 
-        logger.info("Updating run state", run_uid=uid, run_state=run_state)
+        logger.info(
+            "Updating run state - ensure_run_state flow",
```
Comment (TomerShor) [inline server/py/services/api/runtime_handlers/base.py]: What about:

```suggestion
            "Ensuring run state",
```

File: server/py/services/api/tests/unit/runtime_handlers/test_kubejob.py [hunk 2]
```diff
@@ -972,6 +977,33 @@ async def test_retry_pod_deleted_before_first_attempt(
...
+            # first run, no retry label means pod is valid and not outdated
+            (None, 0, False),
+            # retry count > 0 and no retry label is present, pod is outdated
+            (None, 1, True),
+            # pod attempt is older than current run retry, pod is outdated
+            ("1", 2, True),
+            # pod attempt equals current run retry, pod is still valid
+            ("2", 2, False),
+        ],
+    )
```
Comment (TomerShor) [inline server/py/services/api/tests/unit/runtime_handlers/test_kubejob.py]: Is an edge case possible where `pod_retry_label > run_retry_count`?
If so - what do we do in this case?

**Review by TomerShor — CHANGES_REQUESTED** [review]
Looking good!
Minor comments

File: server/py/services/api/runtime_handlers/base.py (hunk 1 above)
Comment (Yacouby) [reply server/py/services/api/runtime_handlers/base.py]: mmm
I think “updating run state” fits better, as we’re actually performing the update, not just ensuring it.

File: server/py/services/api/tests/unit/runtime_handlers/test_kubejob.py (hunk 2 above)
Comment (Yacouby) [reply server/py/services/api/tests/unit/runtime_handlers/test_kubejob.py]: In this edge case, the pod is treated as **not** outdated to avoid skipping a potentially valid, still-running attempt. Added a test case to cover this scenario.

**Review by liranbg — APPROVED** [review]

**Review by TomerShor — APPROVED** [review]
